lib/react_client.dart

```diff
@@ -55,37 +34,53 @@ abstract class ReactComponentFactoryProxy implements Function {
   get type;
 
   /// Returns a new rendered component instance with the specified [props] and [children].
-  JsObject call(Map props, [dynamic children]);
+  ReactElement call(Map props, [dynamic children]);
 
   /// Used to implement a variadic version of [call], in which children may be specified as additional arguments.
   dynamic noSuchMethod(Invocation invocation);
 }
 
+/// Prepares [children] to be passed to ReactJS.
+///
+/// Currently only involves converting a top-level non-[List] [Iterable] to
+/// a non-growable [List]s, but this may be updated in the future to support
+/// advanced nesting and other kinds of children.
+dynamic jsifyChildren(dynamic children) {
+  if (children is Iterable && children is! List) {
+    return children.toList(growable: false);
+  } else {
+    return children;
+  }
+}
+
 /// Creates ReactJS [Component] instances for Dart components.
-class ReactDartComponentFactoryProxy extends ReactComponentFactoryProxy {
-  final JsFunction reactClass;
-  final JsFunction reactComponentFactory;
+class ReactDartComponentFactoryProxy<TComponent extends Component> extends ReactComponentFactoryProxy {
+  final ReactClass reactClass;
+  final Function reactComponentFactory;
+  final Map defaultProps;
 
-  ReactDartComponentFactoryProxy(JsFunction reactClass) :
+  ReactDartComponentFactoryProxy(ReactClass reactClass) :
       this.reactClass = reactClass,
-      this.reactComponentFactory = _React.callMethod('createFactory', [reactClass]);
+      this.reactComponentFactory = React.createFactory(reactClass),
+      this.defaultProps = reactClass.dartDefaultProps
```

Comment:
instead of assigning and immediately asserting, would it make more sense to check for null and perform a reasonable default assignment?
```
this.defaultProps = reactClass.dartDefaultProps ?? {};
```